jubilant/secrettypes.py

```diff
@@ -0,0 +1,146 @@
+"""Dataclasses that contain parsed output from juju secrets."""
+
+from __future__ import annotations
+
+import dataclasses
+import datetime
+import enum
+from typing import Any, Dict, Generic, TypeVar
+
+
+class Rotate(enum.Enum):
+    """Secret rotation policies."""
+
+    NEVER = 'never'
+    HOURLY = 'hourly'
+    DAILY = 'daily'
+    WEEKLY = 'weekly'
+    MONTHLY = 'monthly'
+    QUARTERLY = 'quarterly'
+    YEARLY = 'yearly'
+
+
+class AccessScope(enum.Enum):
+    """Secret access scopes."""
+
+    UNIT = 'unit'
+    APP = 'application'
+    MODEL = 'model'
+    RELATION = 'relation'
+
+
+class AccessRole(enum.Enum):
+    """Secret access roles."""
+
+    VIEW = 'view'
+    ROTATE = 'rotate'
+    MANAGE = 'manage'
+
+
+class SecretURI(str):
+    """A string subclass that represents a secret URI ("secret:...")."""
+
+    @property
+    def unique_identifier(self) -> str:
+        """Unique identifier of this secret URI.
+
+        This is the secret's globally-unique identifier (currently a 20-character Xid,
+        for example "9m4e2mr0ui3e8a215n4g").
+        """
+        if '/' in self:
+            # Handle 'secret://MODEL-UUID/UNIQUE-IDENTIFIER'
+            return self.rsplit('/', maxsplit=1)[-1]
+        elif self.startswith('secret:'):
+            # Handle common case of 'secret:UNIQUE-IDENTIFIER'
+            return self[len('secret:') :]
+        else:
+            return str(self)
+
+
+NoneType = type(None)
+Hidden = NoneType
+Revealed = Dict[str, str]
+T = TypeVar('T', Revealed, Hidden)
+
+
+@dataclasses.dataclass(frozen=True)
+class Secret(Generic[T]):
+    """Represents a secret."""
+
+    uri: SecretURI
+    revision: int
+    checksum: str | None
+    expires: str | None
+    rotation: Rotate | None
+    rotates: datetime.datetime | None
+    owner: str
+    description: str | None
+    name: str | None
+    label: str | None
+    created: datetime.datetime
+    updated: datetime.datetime
+    content: T
+    revisions: list[Revision] | None
+    access: list[Access] | None
+
+    @classmethod
+    def _from_dict(cls, d: dict[str, Any]) -> Secret[T]:
+        return cls(
+            uri=SecretURI('secret:' + d.get('uri', '')),
+            name=d.get('name'),
+            label=d.get('label'),
+            owner=d.get('owner', ''),
+            rotates=datetime.datetime.fromisoformat(d['rotates'].replace('Z', '+00:00'))
+            if 'rotates' in d
+            else None,
```

Comment:
Yeah, so let's add that `_datetime_from_iso` helper and use it here (and elsewhere) like so:

```suggestion
            rotates=_datetime_from_iso(d['rotates']) if 'rotates' in d else None,
```